docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,320 @@
+
+## <a name='Motivation'></a>Motivation
+
+Blocking on success is sometimes the right thing to do but most of the time, it's the equivalent of the UI freezing on an app because you have used the UI thread to fetch some data causing your user to wonder why and when the software will react. This proposal aims to make the reaction time of the CAPZ controller drastically faster and possibly, more resilient.
+
+### <a name='Goals'></a>Goals
+
+- Accelerate the feedback loop with the user so they can know that reconciliation is progressing without having to go check the Azure resources in the portal/CLI/etc.
+- Make the time for the controller to react to a change much faster
+- Improve the resiliency of the controller by making it more fault tolerant
+- Make it easier for the user to understand the state of each resource by adding more granular Conditions
+- Apply the same asynchronous pattern to all resources
+
+### <a name='Non-GoalsFutureWork'></a>Non-Goals / Future Work
+
+- Increase or decrease overall duration of reconciliation
+- Increase the number of API calls to Azure
+- Start Azure operations in parallel
+- Predict how long each operation will take
+
+## <a name='Proposal'></a>Proposal
+
+### <a name='UserStories'></a>User Stories
+
+#### <a name='Story1-UXofcreatinganAzureCluster'></a>Story 1 - UX of creating an AzureCluster
+
+Blake is a Program Manager trying out Cluster API for the first time. Blake is following the quickstart documentation in the Cluster API book and using Azure to create a cluster. Blake applies the cluster template on the management cluster and describes the resulting AzureCluster resource. The AzureCluster is in "Creating" state and the Conditions get updated as Azure resources are created to show the progress.
+
+#### <a name='Story2-ScalingupaMachineDeployment'></a>Story 2 - Creating AzureMachines concurrently
+
+Alex is an engineer in a large organization which has a MachineDeployment running. Alex needs to scale up the number of replicas of the MachineDeployment by 200. Alex uses `kubectl` to scale the number of replicas in the MachineDeployment by 200. The AzureMachine controller in the management cluster is running with the default concurrency of 10. Ten new AzureMachines are created and their state quickly becomes "Creating". Shortly after, before the first ten machines are done creating, ten new ones start creating. The same thing happens until all 200 AzureMachines are in "Creating" state. Alex checks the Conditions on one of the creating AzureMachines and sees that the network interface was created successfully, and that the VM is being created. This allows Alex to quickly scale up the number of replicas as the new 200 VMs get created concurrently, without having to increase the controller concurrency.
```

Comment:
I don't understand this scenario. It seems that with the additional decomposition and asynchronicity of Azure resource reconciliation we are saying that the AzureMachine concurrency configuration is no longer operable?

Reply:
The concurrency limit will still be there and will still apply, but because a single loop will take 1 minute as opposed to 5-10 minutes (average), you can queue up all 200 Azure VM creates in Azure more quickly.

Reply:
Understood